File: test-suite/golden-tests/tensor-comprehensive/src.R
## ---------------------------------------------------------------------------
## Round-trip baselines.
## ---------------------------------------------------------------------------

## Matrix 3x4 Real, row-major values 1..12 at the wire boundary.
## The pack/unpack pair in tensor-r reverses axis order, so after unpack
## a wire row-major fill of 1..12 produces an R matrix whose [i,j] matches
## numpy/Cpp [i-1, j-1] under the row-major semantic indexing.
rMakeMat34 <- function() {
  ## Build a row-major-equivalent 3x4 by filling column-by-column and transposing.
  matrix(as.double(1:12), nrow = 3, ncol = 4, byrow = TRUE)
}

rSumMat34 <- function(m) sum(m)

## Tensor3 2x3x4 Real, row-major values 0..23 at the wire boundary.
## R stores column-major; the unpack converts via aperm so the semantic
## value at index (i,j,k) (zero-based) is i*12 + j*4 + k.
## To produce that semantic content from R, we fill the array with values
## in the order R stores them, then aperm to match the semantic layout.
rMakeT3 <- function() {
  ## We want r[i+1, j+1, k+1] == i*12 + j*4 + k (zero-based semantics).
  ## Easiest: build the row-major sequence then permute axes.
  ## After aperm with perm=c(3,2,1), an R array of dims=c(4,3,2) filled
  ## column-major with seq 0..23 gives an array of dims=c(2,3,4) where
  ## the original linear order maps to the row-major semantics.
  aperm(array(as.double(0:23), dim = c(4, 3, 2)), c(3, 2, 1))
}

rSumT3 <- function(t) sum(t)

## Tensor4 2x3x2x2 Real, row-major values 0..23.
rMakeT4 <- function() {
  aperm(array(as.double(0:23), dim = c(2, 2, 3, 2)), c(4, 3, 2, 1))
}

rSumT4 <- function(t) sum(t)

## Tensor5 2x2x2x2x2 Real, row-major values 0..31.
rMakeT5 <- function() {
  aperm(array(as.double(0:31), dim = c(2, 2, 2, 2, 2)), c(5, 4, 3, 2, 1))
}

rSumT5 <- function(t) sum(t)

## ---------------------------------------------------------------------------
## matmul consumer (Matrix 2 2 Real -> Real).
## ---------------------------------------------------------------------------
rSumMat22 <- function(m) sum(m)

## ---------------------------------------------------------------------------
## Corner extraction. Uses semantic (zero-based) indexing then converts
## to R's one-based. Expected result for the rMakeT3 fixture: c(0,12,11,23).
## ---------------------------------------------------------------------------
rCornersT3 <- function(t) {
  c(t[1, 1, 1],   # (0,0,0)
    t[2, 1, 1],   # (1,0,0)
    t[1, 3, 4],   # (0,2,3)
    t[2, 3, 4])   # (1,2,3)
}

## ---------------------------------------------------------------------------
## Big matrix for SHM-path zero-copy probe.
## ---------------------------------------------------------------------------
rMakeMatBig <- function() {
  matrix(1.0, nrow = 200, ncol = 100)
}

## Polymorphic-in-shape sum for literal-input tests.
rSumMatPoly <- function(m) sum(m)

File: test-suite/golden-tests/tensor-comprehensive/src.py
import numpy as np

# ---------------------------------------------------------------------------
# Round-trip baselines.
# ---------------------------------------------------------------------------
def pyMakeMat34():
    return np.arange(1, 13, dtype=np.float64).reshape(3, 4)

def pySumMat34(m):
    return float(np.sum(m))

def pyMakeT3():
    return np.arange(24, dtype=np.float64).reshape(2, 3, 4)

def pySumT3(t):
    return float(np.sum(t))

def pyMakeT4():
    return np.arange(24, dtype=np.float64).reshape(2, 3, 2, 2)

def pySumT4(t):
    return float(np.sum(t))

def pyMakeT5():
    return np.arange(32, dtype=np.float64).reshape(2, 2, 2, 2, 2)

def pySumT5(t):
    return float(np.sum(t))

# ---------------------------------------------------------------------------
# matmul inputs and consumer.
# ---------------------------------------------------------------------------
def pyMakeA23():
    return np.arange(1, 7, dtype=np.float64).reshape(2, 3)

def pyMakeB32():
    return np.arange(1, 7, dtype=np.float64).reshape(3, 2)

def pySumMat22(m):
    return float(np.sum(m))

# ---------------------------------------------------------------------------
# Corner extraction probe. Indices use the semantic (row-major) view.
# numpy is row-major by default, so t[i,j,k] matches Cpp t(i,j,k).
# Expected: [0, 12, 11, 23] for a Tensor3 2x3x4 with values 0..23.
# ---------------------------------------------------------------------------
def pyCornersT3(t):
    return [float(t[0,0,0]), float(t[1,0,0]), float(t[0,2,3]), float(t[1,2,3])]

# ---------------------------------------------------------------------------
# Element-typed matrices (4x4, values 1..16, sum = 136).
# ---------------------------------------------------------------------------
def pyMakeMatI32():
    return np.arange(1, 17, dtype=np.int32).reshape(4, 4)

def pySumMatI32(m):
    return int(np.sum(m.astype(np.int64)))

def pyMakeMatI64():
    return np.arange(1, 17, dtype=np.int64).reshape(4, 4)

def pySumMatI64(m):
    return int(sum(int(x) for x in m.reshape(-1).tolist()))

def pyMakeMatF32():
    return np.arange(1, 17, dtype=np.float32).reshape(4, 4)

def pySumMatF32(m):
    return float(np.sum(m))

# ---------------------------------------------------------------------------
# Zero-copy probes. The user-facing ndarray for ranks >= 2 is the result
# of reshape(), which always returns a view: DIRECT (its own OWNDATA) is
# always False. The witness for SHM zero-copy is the OWNDATA of the ROOT
# of the .base chain. If ROOT=False, the buffer is external (SHM-backed
# numpy view); if ROOT=True, somewhere up the chain a numpy-owned buffer
# was allocated (a copy happened).
# ---------------------------------------------------------------------------
def _ownroot(arr):
    while hasattr(arr, 'base') and arr.base is not None:
        arr = arr.base
    if hasattr(arr, 'flags'):
        return bool(arr.flags['OWNDATA'])
    return None

def _report(arr):
    direct = bool(arr.flags['OWNDATA']) if hasattr(arr, 'flags') else None
    root = _ownroot(arr)
    size = int(arr.size if hasattr(arr, 'size') else 0)
    return "DIRECT=%s,ROOT=%s,size=%d" % (str(direct), str(root), size)

def pyOwnDataMat(m):
    return _report(m)

def pyOwnDataT3(t):
    return _report(t)

def pyOwnDataT5(t):
    return _report(t)

# ---------------------------------------------------------------------------
# Edge cases.
# ---------------------------------------------------------------------------
def pyMakeUnitMat():
    return np.array([[7.0]], dtype=np.float64)

def pySumUnitMat(m):
    return float(np.sum(m))

def pyMakeUnitT3():
    return np.array([[[9.0]]], dtype=np.float64)

def pySumUnitT3(t):
    return float(np.sum(t))

# ---------------------------------------------------------------------------
# Polymorphic-in-shape sum helpers for literal-input tests. The morloc
# Matrix m n a / Tensor3 d1 d2 d3 a maps to numpy.ndarray regardless of
# specific dims, so these accept any shape.
# ---------------------------------------------------------------------------
def pySumMatPoly(m):
    return float(np.sum(m))

def pySumMatI32Poly(m):
    return int(np.sum(m.astype(np.int64) if hasattr(m, 'astype') else m))

def pySumT3Poly(t):
    return float(np.sum(t))
